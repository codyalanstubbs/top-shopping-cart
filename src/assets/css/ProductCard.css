p {
    margin: 0;
    text-align: center;
}

.sample {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 10px;
    align-items: center;
    border: 2px solid #3882F6;
    border-radius: 8px;
    padding: 10px;
}

.sample:hover {
    background-color: #adccff;
}

.p-qty-container {
    display: flex;
    gap: 10px;
}

.qty-adjuster {
    text-align: center;
    height: 20px;
    width: 20px;
    border-radius: 4px;
    border: 1px solid grey;
}

.qty-adjuster:hover {
    background-color: #3882F6;
    border: 1px solid #adccff;;
}