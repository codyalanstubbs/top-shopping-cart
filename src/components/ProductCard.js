import React, { useState } from "react";
import FlavorWheelImage from "../assets/images/flavor-wheel.png";
import "../assets/css/ProductCard.css";
const ProductCard = ({productData, increaseQtyInCart, decreaseQtyInCart, setQtyInCart}) => {
    const {id, name, cost, qty} = productData;

    const handleIncreaseQty = (event) => {
        const productID = event.target.id;
        const qtyInput = document.querySelector(`#\\3${productID}.qty`);

        if (qtyInput.value >= 0) {
            const newValue = Number(qtyInput.value) + 1;
            qtyInput.value = newValue;
        }
    };

    const handleDecreaseQty = (event) => {
        const productID = event.target.id;
        const qtyInput = document.querySelector(`#\\3${productID}.qty`);

        if (qtyInput.value > 0) {
            const newValue = Number(qtyInput.value) - 1;
            qtyInput.value = newValue;
        }
    };

    const onTextChange = (event) => {
        const newValue = event.target.value;
        if (newValue >= 0) {
            event.target.value = newValue;
        }
    }

    const changeProductQuantity = (event) => {
        const productID = event.target.id;
        const qtyInput = document.querySelector(`#\\3${productID}.qty`);

        if (qtyInput.value >= 0) {
            setQtyInCart(qtyInput.value, productID);
            qtyInput.value = 0;
        }
    }
    
    return (
        <div
            key={id}
            id={id}
            data-testid="p-card" 
            className="sample"
        >
            <img src={FlavorWheelImage} alt="A flavor wheel for green tea." width="100px" height="100px" />
            <p data-testid="p-name">{name} Lexicon</p>
            <p data-testid="p-cost">{"$"+cost}</p>
            <div className="p-qty-container">
                <button id={id} className="qty-adjuster" onClick={handleDecreaseQty} >-</button>
                <input id={id} className="qty" placeholder="0" onChange={onTextChange} />
                <button id={id} className="qty-adjuster" onClick={handleIncreaseQty} >+</button>
            </div>
            <button id={id} className="add" onClick={changeProductQuantity}>ADD</button>
        </div>
    );       
};

export default ProductCard;