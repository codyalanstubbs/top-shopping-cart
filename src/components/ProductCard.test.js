import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from "@testing-library/user-event";
import ProductCard from './ProductCard';

const defaultProductData = {
    id: "Ex2ods",
    name: "Example Product",
    cost: 100,
    qty: 0
}

const renderProductCardComponent = (productData) => render(
        <ProductCard
            productData={productData}
            addToCart={jest.fn()}
            handleIncreaseQnt={jest.fn()}
            handleDecreaseQnt={jest.fn()}
        />
);

describe('Product Card', () => {

    describe('Render default ProductCard', () => {
        test("Renders ProductCard", () => {
            const utils = renderProductCardComponent({...defaultProductData});
            expect(utils).toMatchSnapshot();
        });
    });
    
    describe('Tests for "static" product information', () => {
        test('Renders product card with id of "XUAk3"', () => {

            renderProductCardComponent({
                ...defaultProductData,
                id: 'XUAk3',
            });
    
            const productCard = screen.getByTestId('p-card');
            expect(productCard).toBeInTheDocument();         
            expect(productCard).toHaveAttribute('id','XUAk3');
            expect(productCard).not.toHaveAttribute('id', defaultProductData.id);

        });

        test('Renders product card with name of "Apple Lexicon"', () => {
 
             renderProductCardComponent({
                 ...defaultProductData,
                 name: 'Apple',
             });
     
             const productName = screen.getByText('Apple Lexicon');
             expect(productName).toBeInTheDocument();            
 
        });

        test('Renders product card with p-cost that has text of "$500"', () => {
 
             renderProductCardComponent({
                 ...defaultProductData,
                 cost: 500,
             });
     
             const productCost = screen.getByTestId('p-cost');
             expect(productCost.textContent).toBe("$500");
             expect(productCost.textContent).not.toBe("500");
             expect(productCost.textContent).not.toBe(500);  
             expect(productCost.textContent).not.toBe(defaultProductData.cost);
             expect(productCost.textContent).not.toBe("$"+defaultProductData.cost);
 
        });
    });

    describe("Test dynamic, UI-driven changes to product data", () => {

        test("Increases product quantity by 1", () => {
            renderProductCardComponent({...defaultProductData});
            const increaseQuantityBtn = screen.getByText('+');
            expect(increaseQuantityBtn.className).toBe("qty-adjuster");

            userEvent.click(increaseQuantityBtn);

            const productQuantity = screen.getByTestId("p-qty");
            expect(productQuantity).toBeInTheDocument();
            expect(productQuantity.textContent).toBe("1");
        });

        test("Does not decrease product quantity when already 0", () => {
            renderProductCardComponent({...defaultProductData, qty: 0});
            const decreaseQuantityBtn = screen.getByText('-');
            expect(decreaseQuantityBtn.className).toBe("qty-adjuster");

            userEvent.click(decreaseQuantityBtn);

            const productQuantity = screen.getByTestId("p-qty");
            expect(productQuantity).toBeInTheDocument();
            expect(productQuantity.textContent).toBe("0");
            expect(productQuantity.textContent).not.toBe("-1");
        });

        test("Decreases product quantity by 1 when quantity > 0", () => {
            renderProductCardComponent({...defaultProductData, qty: 1});
            const decreaseQuantityBtn = screen.getByText('-');
            expect(decreaseQuantityBtn.className).toBe("qty-adjuster");

            userEvent.click(decreaseQuantityBtn);

            const productQuantity = screen.getByTestId("p-qty");
            expect(productQuantity).toBeInTheDocument();
            expect(productQuantity.textContent).toBe("0");
        });

      });
    

});
