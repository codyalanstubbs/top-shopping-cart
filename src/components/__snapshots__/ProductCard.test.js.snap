// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Render default ProductCard Renders ProductCard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sample"
        data-testid="p-card"
        id="Ex2ods"
      >
        <img
          alt="A flavor wheel for green tea."
          height="100px"
          src="flavor-wheel.png"
          width="100px"
        />
        <p
          data-testid="p-name"
        >
          Example Product
           Lexicon
        </p>
        <p
          data-testid="p-cost"
        >
          $100
        </p>
        <div
          class="p-qty-container"
        >
          <button
            class="qty-adjuster"
            id="Ex2ods"
          >
            -
          </button>
          <input
            class="qty"
            id="Ex2ods"
            placeholder="0"
          />
          <button
            class="qty-adjuster"
            id="Ex2ods"
          >
            +
          </button>
        </div>
        <button
          class="add"
          id="Ex2ods"
        >
          ADD
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sample"
      data-testid="p-card"
      id="Ex2ods"
    >
      <img
        alt="A flavor wheel for green tea."
        height="100px"
        src="flavor-wheel.png"
        width="100px"
      />
      <p
        data-testid="p-name"
      >
        Example Product
         Lexicon
      </p>
      <p
        data-testid="p-cost"
      >
        $100
      </p>
      <div
        class="p-qty-container"
      >
        <button
          class="qty-adjuster"
          id="Ex2ods"
        >
          -
        </button>
        <input
          class="qty"
          id="Ex2ods"
          placeholder="0"
        />
        <button
          class="qty-adjuster"
          id="Ex2ods"
        >
          +
        </button>
      </div>
      <button
        class="add"
        id="Ex2ods"
      >
        ADD
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
